src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs

```diff
@@ -0,0 +1,209 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using OrchardCore.ContentManagement;
+using OrchardCore.Entities;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Scope;
+using OrchardCore.Indexing;
+using OrchardCore.Modules;
+using OrchardCore.Settings;
+
+namespace OrchardCore.Lucene
+{
+    /// <summary>
+    /// This class provides services to update all the Lucene indices. It is non-rentrant so that calls 
+    /// from different components can be done simultaneously, e.g. from a background task, an event or a UI interaction.
+    /// It also indexes one content item at a time and provides the result to all indices.
+    /// </summary>
+    public class LuceneIndexingService
+    {
+        private const int BatchSize = 100;
+        private readonly IShellHost _shellHost;
+        private readonly ShellSettings _shellSettings;
+        private readonly LuceneIndexingState _indexingState;
+        private readonly LuceneIndexSettings _luceneIndexSettings;
+        private readonly LuceneIndexManager _indexManager;
+        private readonly IIndexingTaskManager _indexingTaskManager;
+        private readonly ISiteService _siteService;
+
+        public LuceneIndexingService(
+            IShellHost shellHost,
+            ShellSettings shellSettings,
+            LuceneIndexingState indexingState,
+            LuceneIndexSettings luceneIndexSettings,
+            LuceneIndexManager indexManager,
+            IIndexingTaskManager indexingTaskManager,
+            ISiteService siteService,
+            ILogger<LuceneIndexingService> logger)
+        {
+            _shellHost = shellHost;
+            _shellSettings = shellSettings;
+            _indexingState = indexingState;
+            _luceneIndexSettings = luceneIndexSettings;
+            _indexManager = indexManager;
+            _indexingTaskManager = indexingTaskManager;
+            _siteService = siteService;
+
+            Logger = logger;
+        }
+
+        public ILogger Logger { get; }
+
+        public async Task ProcessContentItemsAsync(string indexName = default)
```

Comment:
This was not processing one index at a time. I kept the option but now we can pass the indexName to start processing a specific index.

Reply:
Hmm, we would need to check again how indexing sates are working (e.g last task ids in `lucene.status.json`), but if only one index has been rebuilt, not sure that it was doing extra work related to the other ones.

Reply:
That's what I experienced. I need to retry, maybe it was because of the Lucene Background worker. But you can see in the code that it loops on all indexes.